include/seqan3/contrib/stream/bgzf_istream.hpp

```diff
@@ -0,0 +1,635 @@
+// zipstream Library License:
+// --------------------------
+//
+// The zlib/libpng License Copyright (c) 2003 Jonathan de Halleux.
+//
+// This software is provided 'as-is', without any express or implied warranty. In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose, including commercial applications, and to alter it and redistribute it freely, subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented; you must not claim that you wrote the original software. If you use this software in a product, an acknowledgment in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such, and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution
+//
+//
+// Author: Jonathan de Halleux, dehalleux@pelikhan.com, 2003   (original zlib stream)
+// Author: David Weese, dave.weese@gmail.com, 2014             (extension to parallel block-wise compression in bgzf format)
+// Author: Rene Rahn, rene.rahn AT fu-berlin.de, 2019          (adaption to SeqAn3)
+
+#pragma once
+
+// TODO(rrahn): replace me with constructor argument
+#ifndef SEQAN_BGZF_NUM_THREADS
+    #define SEQAN_BGZF_NUM_THREADS 16
+#endif
+
+#include <cstring>
+#include <condition_variable>
+#include <mutex>
+
+#include <seqan3/contrib/parallel/buffer_queue.hpp>
+#include <seqan3/contrib/stream/bgzf_stream_util.hpp>
+#include <seqan3/core/parallel/detail/reader_writer_manager.hpp>
+#include <seqan3/io/exception.hpp>
+
+namespace seqan3::contrib
+{
+
+const unsigned BGZF_MAX_BLOCK_SIZE = 64 * 1024;
+const unsigned BGZF_BLOCK_HEADER_LENGTH = 18;
+const unsigned BGZF_BLOCK_FOOTER_LENGTH = 8;
+const unsigned ZLIB_BLOCK_OVERHEAD = 5; // 5 bytes block overhead (see 3.2.4. at http://www.gzip.org/zlib/rfc-deflate.html)
+
+// Reduce the maximal input size, such that the compressed data
+// always fits in one block even for level Z_NO_COMPRESSION.
+const unsigned BGZF_BLOCK_SIZE = BGZF_MAX_BLOCK_SIZE - BGZF_BLOCK_HEADER_LENGTH - BGZF_BLOCK_FOOTER_LENGTH - ZLIB_BLOCK_OVERHEAD;
+
+// ===========================================================================
+// Classes
+// ===========================================================================
+
+// --------------------------------------------------------------------------
+// Class basic_decompression_bgzf_streambuf
+// --------------------------------------------------------------------------
+
+template<
+    typename Elem,
+    typename Tr = std::char_traits<Elem>,
+    typename ElemA = std::allocator<Elem>,
+    typename ByteT = char,
+    typename ByteAT = std::allocator<ByteT>
+>
+class basic_decompression_bgzf_streambuf : public std::basic_streambuf<Elem, Tr>
```

Comment:
I would recommend to make the naming similar to the other streams so it is easier to interface from SeqAn, i.e. call this class `basic_bgzf_istreambuf` (for gz it's called `basic_gz_istreambuf`  and for bz2 it's called `basic_bz2_istreambuf`).